spec/factories.rb

```diff
@@ -0,0 +1,55 @@
+# Copyright 2018 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FactoryBot.define do
+  # Using YAML parsing short circuits the object lifecycle.
```

Comment:
What is the normal object lifecycle? How do we short circuit it?

Reply:
Here's how one would create a normal object:

```
  obj = Object.new(some_args)
```

(We don't really do that anywhere in the code)

What happens under the hood? This is my best understanding (written in pseudo-ish code).

```
  obj = allocate_some_memory
  obj.set_class = Object
  obj.initialize(some_args)
```

That's why we call `.new()` to create a new object, but the constructor functions are `.initialize`

Now, how do we create objects? We deserialize them directly from YAML. `.new` never gets called. `.initialize` never gets called. We just immediately get an object of whatever class we specify in YAML with whatever fields are specified in that YAML file. You can add fields in YAML that don't actually exist on the class and they'll end up on the object!